README.md
# NAME

LCS - Longest Common Subsequence

<div>
    <a href="https://travis-ci.org/wollmers/LCS"><img src="https://travis-ci.org/wollmers/LCS.png" alt="LCS"></a>
    <a href='https://coveralls.io/r/wollmers/LCS?branch=master'><img src='https://coveralls.io/repos/wollmers/LCS/badge.png?branch=master' alt='Coverage Status' /></a>
    <a href='http://cpants.cpanauthors.org/dist/LCS'><img src='http://cpants.cpanauthors.org/dist/LCS.png' alt='Kwalitee Score' /></a>
    <a href="http://badge.fury.io/pl/LCS"><img src="https://badge.fury.io/pl/LCS.svg" alt="CPAN version" height="18"></a>
</div>

# SYNOPSIS

    use LCS;
    my $lcs = LCS->LCS( [qw(a b)], [qw(a b b)] );

    # $lcs now contains an arrayref of matching positions
    # same as
    $lcs = [
      [ 0, 0 ],
      [ 1, 2 ]
    ];

    my $all_lcs = LCS->allLCS( [qw(a b)], [qw(a b b)] );

    # same as
    $all_lcs = [
      [
        [ 0, 0 ],
        [ 1, 1 ]
      ],
      [
        [ 0, 0 ],
        [ 1, 2 ]
      ]
    ];

# DESCRIPTION

LCS is an implementation based on the traditional LCS algorithm.

It contains reference implementations working slow but correct.

Also some utility methods are added to reformat the result.

## CONSTRUCTOR

- new()

    Creates a new object which maintains internal storage areas
    for the LCS computation.  Use one of these per concurrent
    LCS() call.

## METHODS

- LCS(\\@a,\\@b)

    Finds a Longest Common Subsequence, taking two arrayrefs as method
    arguments. It returns an array reference of corresponding
    indices, which are represented by 2-element array refs.

        # position  0 1 2
        my $a = [qw(a b  )];
        my $b = [qw(a b b)];

        my $lcs = LCS->LCS($a,$b);

        # same like
        $lcs = [
            [ 0, 0 ],
            [ 1, 1 ]
        ];

- LLCS(\\@a,\\@b)

    Calculates the length of the Longest Common Subsequence.

        my $llcs = LCS->LLCS( [qw(a b)], [qw(a b b)] );
        print $llcs,"\n";   # prints 2

        # is the same as
        $llcs = scalar @{LCS->LCS( [qw(a b)], [qw(a b b)] )};

- allLCS(\\@a,\\@b)

    Finds all Longest Common Subsequences. It returns an array reference of all
    LCS.

        my $all_lcs = LCS->allLCS( [qw(a b)], [qw(a b b)] );

        # same as
        $all_lcs = [
          [
            [ 0, 0 ],
            [ 1, 1 ]
          ],
          [
            [ 0, 0 ],
            [ 1, 2 ]
          ]
        ];

    The purpose is mainly for testing LCS algorithms, as they only return one of the optimal
    solutions. If we want to know, that the result is one of the optimal solutions, we need
    to test, if the solution is part of all optimal LCS:

        use Test::More;
        use Test::Deep;
        use LCS;
        use LCS::Tiny;

        cmp_deeply(
          LCS::Tiny->LCS(\@a,\@b),
          any(@{LCS->allLCS(\@a,\@b)} ),
          "Tiny::LCS $a, $b"
        );

- lcs2align(\\@a,\\@b,$LCS)

    Returns the two sequences aligned, missing positions are represented as empty strings.

        use Data::Dumper;
        use LCS;
        print Dumper(
          LCS->lcs2align(
            [qw(a   b)],
            [qw(a b b)],
            LCS->LCS([qw(a b)],[qw(a b b)])
          )
        );
        # prints

        $VAR1 = [
                  [
                    'a',
                    'a'
                  ],
                  [
                    '',
                    'b'
                  ],
                  [
                    'b',
                    'b'
                  ]
        ];

- align(\\@a,\\@b)

    Returns the same as lcs2align() via calling LCS() itself.

- sequences2hunks($a, $b)

    Transforms two array references of scalars to an array of hunks (two element arrays).

- hunks2sequences($hunks)

    Transforms an array of hunks to two arrays of scalars.

        use Data::Dumper;
        use LCS;
        print Dumper(
          LCS->hunks2sequences(
            LCS->LCS([qw(a b)],[qw(a b b)])
          )
        );
        # prints (reformatted)
        $VAR1 = [ 0, 1 ];
        $VAR2 = [ 0, 2 ];

- align2strings($hunks, $gap\_character)

    Returns two strings aligned with gap characters. The default gap character is '\_'.

        use Data::Dumper;
        use LCS;
        print Dumper(
          LCS->align2strings(
            LCS->lcs2align([qw(a b)],[qw(a b b)],LCS->LCS([qw(a b)],[qw(a b b)]))
          )
        );
        $VAR1 = 'a_b';
        $VAR2 = 'abb';

- fill\_strings($string1, $string2, $fill\_character)

    Returns both strings filling up the shorter with $fill\_character to the same length.

    The default $fill\_character is '\_'.

- clcs2lcs($compact\_lcs)

    Convert compact LCS to LCS.

- lcs2clcs($compact\_lcs)

    Convert LCS to compact LCS.

- max($i, $j)

    Returns the maximum of two numbers.

## EXPORT

None by design.

# STABILITY

Until release of version 1.00 the included methods, names of methods and their
interfaces are subject to change.

Beginning with version 1.00 the specification will be stable, i.e. not changed between
major versions.

# REFERENCES

Ronald I. Greenberg. Fast and Simple Computation of All Longest Common Subsequences,
http://arxiv.org/pdf/cs/0211001.pdf

Robert A. Wagner and Michael J. Fischer. The string-to-string correction problem.
Journal of the ACM, 21(1):168-173, 1974.

# SOURCE REPOSITORY

[http://github.com/wollmers/LCS](http://github.com/wollmers/LCS)

# AUTHOR

Helmut Wollmersdorfer &lt;helmut.wollmersdorfer@gmail.com>

<div>
    <a href='http://cpants.cpanauthors.org/author/wollmers'><img src='http://cpants.cpanauthors.org/author/wollmers.png' alt='Kwalitee Score' /></a>
</div>

# COPYRIGHT

Copyright 2014- Helmut Wollmersdorfer

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO


Changes
Revision history for LCS

{{$NEXT}}

0.11  2016-11-25 08:50:35 CET
        - added build requires, MetaProvides::Package

0.10  2015-06-26 05:20:55 CEST
        - improved documentation

0.09  2015-06-01 13:55:25 CEST
        - stable release

0.08  2015-05-28 11:25:16 CEST
        - added author tests

0.07  2015-05-27 20:29:07 CEST
        - added spelling test

0.06  2015-05-26 06:36:32 CEST
        - release

0.05  2015-05-25 21:52:40 CEST
        - improved backwards compatibility

0.04  2015-05-19 22:55:36 CEST
        - downgrade minimum Perl version

0.03  2015-05-09 11:17:41 CEST
        - added benchmarks
        - loopified recursions

0.02  2015-05-06 07:10:53 CEST
        - added tests

0.01  2015-05-03 20:30:11 CEST
        - Initial release


lib/LCS/Tiny.pm
package LCS::Tiny;

use 5.006;
use strict;
use warnings;
our $VERSION = '0.10';
#use utf8;

sub new {
  my $class = shift;
  # uncoverable condition false
  bless @_ ? @_ > 1 ? {@_} : {%{$_[0]}} : {}, ref $class || $class;
}

# Implemented according to

# J. W. Hunt and T. G. Szymanski.
# A fast algorithm for computing longest common subsequences.
# Commun. ACM, 20(5):350-353, 1977.

sub LCS {
  my ($self, $a, $b) = @_;

  my ($amin, $amax, $bmin, $bmax) = (0, $#$a, 0, $#$b);

  while ($amin <= $amax and $bmin <= $bmax and $a->[$amin] eq $b->[$bmin]) {
    $amin++;
    $bmin++;
  }
  while ($amin <= $amax and $bmin <= $bmax and $a->[$amax] eq $b->[$bmax]) {
    $amax--;
    $bmax--;
  }

  my $bMatches;
  unshift @{ $bMatches->{$b->[$_]} },$_ for $bmin..$bmax;

  my $matchVector = [];
  my $thresh = [];
  my $links  = [];

  my ( $i, $ai, $j, $k );
  for ( $i = $amin ; $i <= $amax ; $i++ ) {
    $ai = $a->[$i];
    if ( exists( $bMatches->{$ai} ) ) {
      for $j ( @{ $bMatches->{$ai} } ) {
        if ( !@$thresh || $j > $thresh->[-1] ) {
          $k = $#$thresh+1;
          $thresh->[$k] = $j;
        }
        #elsif ( $k and $thresh->[$k] > $j and $thresh->[ $k - 1 ] < $j ) {
        #  $thresh->[$k] = $j;
        #}
        else {
          # binary search for insertion point
          $k = 0;
          my $index;
          my $found;
          my $high = $#$thresh;
          while ( $k <= $high ) {
            use integer;
            $index = ( $high + $k ) / 2;
            #$index = int(( $high + $k ) / 2);  # without 'use integer'
            $found = $thresh->[$index];

            if ( $j == $found ) { $k = undef; last; }
            elsif ( $j > $found ) { $k = $index + 1; }
            else { $high = $index - 1; }
          }
          # now insertion point is in $k.
          $thresh->[$k] = $j if (defined $k);    # overwrite next larger
        }
        if (defined $k) {
          $links->[$k] = [ ( $k ? $links->[ $k - 1 ] : undef ), $i, $j ];
        }
      }
    }
  }
  if (@$thresh) {
    for ( my $link = $links->[$#$thresh] ; $link ; $link = $link->[0] ) {
      unshift @$matchVector,[$link->[1],$link->[2]];
    }
  }
  return [ map([$_ => $_], 0 .. ($bmin-1)),
        @$matchVector,
            map([++$amax => $_], ($bmax+1) .. $#$b) ];
}

1;

__END__

=head1 NAME

LCS::Tiny - Tiny implementation of the
                 Longest Common Subsequence (LCS) Algorithm

=head1 SYNOPSIS

  use LCS::Tiny;

  $alg = LCS::Tiny->new;
  $lcs = $alg->LCS(\@a,\@b);

=head1 ABSTRACT

LCS::Tiny is a heavily tuned version based on Algorithm::Diff

=head1 DESCRIPTION

=head2 CONSTRUCTOR

=over 4

=item new()

Creates a new object which maintains internal storage areas
for the LCS computation.  Use one of these per concurrent
LCS() call.

=back

=head2 METHODS

=over 4


=item LCS(\@a,\@b)

Finds a Longest Common Subsequence, taking two arrayrefs as method
arguments. It returns a list of corresponding
indices, which are represented by 2-element array refs.

=back

=head2 EXPORT

None by design.

=head1 SEE ALSO

Algorithm::Diff

=head1 REFERENCES

James W. Hunt and Thomas G. Szymanski. A fast algorithm for computing longest common
subsequences. Communications of the ACM, 20(5):350-353, 1977.

Hunt, J.W. and McIlroy, M.D. An Algorithm for Differential File Comparison.
Computing Science Technical Report 41, Bell Laboratories (1975).

=head1 AUTHOR

Helmut Wollmersdorfer E<lt>helmut.wollmersdorfer@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2014-2015 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/LCS.pm
package LCS;

use strict;
use warnings;

use 5.006;
our $VERSION = '0.10';

use Data::Dumper;

sub new {
  my $class = shift;
  # uncoverable condition false
  bless @_ ? @_ > 1 ? {@_} : {%{$_[0]}} : {}, ref $class || $class;
}

sub align {
  my ($self, $X, $Y) = @_;

  return $self->lcs2align(
    $X, $Y, $self->LCS($X,$Y)
  );
}

sub lcs2align {
  my ($self, $X, $Y, $LCS) = @_;

  my $hunks = [];

  my $Xcurrent = -1;
  my $Ycurrent = -1;
  my $Xtemp;
  my $Ytemp;

  for my $hunk (@$LCS) {
    while ( ($Xcurrent+1 < $hunk->[0] ||  $Ycurrent+1 < $hunk->[1]) ) {
      $Xtemp = '';
      $Ytemp = '';
      if ($Xcurrent+1 < $hunk->[0]) {
        $Xcurrent++;
        $Xtemp = $X->[$Xcurrent];
      }
      if ($Ycurrent+1 < $hunk->[1]) {
        $Ycurrent++;
        $Ytemp = $Y->[$Ycurrent];
      }
      push @$hunks,[$Xtemp,$Ytemp];
    }

    $Xcurrent = $hunk->[0];
    $Ycurrent = $hunk->[1];
    push @$hunks,[$X->[$Xcurrent],$Y->[$Ycurrent]]; # elements
  }
  while ( ($Xcurrent+1 <= $#$X ||  $Ycurrent+1 <= $#$Y) ) {
    $Xtemp = '';
    $Ytemp = '';
    if ($Xcurrent+1 <= $#$X) {
      $Xcurrent++;
      $Xtemp = $X->[$Xcurrent];
    }
    if ($Ycurrent+1 <= $#$Y) {
      $Ycurrent++;
      $Ytemp = $Y->[$Ycurrent];
    }
    push @$hunks,[$Xtemp,$Ytemp];
  }
  return $hunks;
}

sub sequences2hunks {
  my ($self, $a, $b) = @_;
  return [ map { [ $a->[$_], $b->[$_] ] } 0..$#$a ];
}

sub clcs2lcs {
  my ($self, $clcs) = @_;
  my $lcs = [];
  for my $entry (@$clcs) {
    for (my $i = 0; $i < $entry->[2];$i++) {
      push @$lcs,[$entry->[0]+$i,$entry->[1]+$i];
    }
  }
  return $lcs;
}

sub lcs2clcs {
  my ($self, $lcs) = @_;
  my $clcs = [];
  for my $entry (@$lcs) {
    if (@$clcs && $clcs->[-1]->[0] + $clcs->[-1]->[2] == $entry->[0]) {
      $clcs->[-1]->[2]++;
    }
    else {
      push @$clcs,[$entry->[0],$entry->[1],1];
    }
  }
  return $clcs;
}

sub hunks2sequences {
  my ($self, $hunks) = @_;

  my $a = [];
  my $b = [];

  for my $hunk (@$hunks) {
    push @$a, $hunk->[0];
    push @$b, $hunk->[1];
  }
  return ($a,$b);
}

sub align2strings {
  my ($self, $hunks,$gap) = @_;
  #$gap //= '_';
  $gap = (defined $gap) ? $gap : '_';

  my $a = '';
  my $b = '';

  for my $hunk (@$hunks) {
    my ($ae,$be) = $self->fill_strings($hunk->[0],$hunk->[1],$gap);
    $a .=  $ae;
    $b .=  $be;
  }
  return ($a,$b);
}

sub fill_strings {
  my ($self, $string1,$string2, $gap) = @_;
  #$gap //= '_';
  $gap = (defined $gap) ? $gap : '_';

  my @m = $string1 =~ m/(\X)/g;
  my @n = $string2 =~ m/(\X)/g;
  my $max = max(scalar(@m),scalar(@n));
  if ($max - scalar(@m) > 0) {
    for (1..$max-scalar(@m)) {
      $string1 .= $gap;
    }
  }
  if ($max - scalar(@n) > 0) {
    for (1..$max-scalar(@n)) {
      $string2 .= $gap;
    }
  }
  return ($string1,$string2);
}

sub LLCS {
  my ($self,$X,$Y) = @_;

  my $m = scalar @$X;
  my $n = scalar @$Y;

  my $c = [];

  for my $i (0..1) {
    for my $j (0..$n) {
      $c->[$i][$j]=0;
    }
  }

  my ($i,$j);

  for ($i=1; $i <= $m; $i++) {
    for ($j=1; $j <= $n; $j++) {
      if ($X->[$i-1] eq $Y->[$j-1]) {
        $c->[1][$j] = $c->[0][$j-1]+1;
      }
      else {
        $c->[1][$j] = max($c->[1][$j-1],$c->[0][$j]);
      }
    }
    for ($j = 1; $j <= $n; $j++) {
      $c->[0][$j] = $c->[1][$j];
    }
  }
  return ($c->[1][$n]);
}


sub LCS {
  my ($self,$X,$Y) = @_;

  my $m = scalar @$X;
  my $n = scalar @$Y;

  my $c = [];
  my ($i,$j);
  for ($i=0;$i<=$m;$i++) {
    for ($j=0;$j<=$n;$j++) {
      $c->[$i][$j]=0;
    }
  }
  for ($i=1;$i<=$m;$i++) {
    for ($j=1;$j<=$n;$j++) {
      if ($X->[$i-1] eq $Y->[$j-1]) {
        $c->[$i][$j] = $c->[$i-1][$j-1]+1;
      }
      else {
        $c->[$i][$j] = max($c->[$i][$j-1], $c->[$i-1][$j]);
      }
    }
  }
  my $path = $self->_lcs($X,$Y,$c,$m,$n,[]);

  return $path;
}


sub max {
  ($_[0] > $_[1]) ? $_[0] : $_[1];
}


sub _lcs {
  my ($self,$X,$Y,$c,$i,$j,$L) = @_;

  while ($i > 0 && $j > 0) {
    if ($X->[$i-1] eq $Y->[$j-1]) {
      unshift @{$L},[$i-1,$j-1];
      $i--;
      $j--;
    }
    elsif ($c->[$i][$j] == $c->[$i-1][$j]) {
      $i--;
    }
    else {
      $j--;
    }
  }
  return $L;
}


sub _all_lcs {
  my ($self,$ranks,$rank,$max) = @_;

  my $R = [[]];

  while ($rank <= $max) {
    my @temp;
    for my $path (@$R) {
      for my $hunk (@{$ranks->{$rank}}) {
        if (scalar @{$path} == 0) {
          push @temp,[$hunk];
        }
        elsif (($path->[-1][0] < $hunk->[0]) && ($path->[-1][1] < $hunk->[1])) {
          push @temp,[@$path,$hunk];
        }
      }
    }
    @$R = @temp;
    $rank++;
  }
  return $R;
}

# get all LCS of two arrays
# records the matches by rank
sub allLCS {
  my ($self,$X,$Y) = @_;

  my $m = scalar @$X;
  my $n = scalar @$Y;

  my $ranks = {}; # e.g. '4' => [[3,6],[4,5]]
  my $c = [];
  my ($i,$j);

  for (0..$m) {$c->[$_][0]=0;}
  for (0..$n) {$c->[0][$_]=0;}
  for ($i=1;$i<=$m;$i++) {
    for ($j=1;$j<=$n;$j++) {
      if ($X->[$i-1] eq $Y->[$j-1]) {
        $c->[$i][$j] = $c->[$i-1][$j-1]+1;
        push @{$ranks->{$c->[$i][$j]}},[$i-1,$j-1];
      }
      else {
        $c->[$i][$j] =
          ($c->[$i][$j-1] > $c->[$i-1][$j])
            ? $c->[$i][$j-1]
            : $c->[$i-1][$j];
      }
    }
  }
  my $max = scalar keys %$ranks;
  return $self->_all_lcs($ranks,1,$max);
}

1;
__END__

=encoding utf-8

=head1 NAME

LCS - Longest Common Subsequence

=begin html

<a href="https://travis-ci.org/wollmers/LCS"><img src="https://travis-ci.org/wollmers/LCS.png" alt="LCS"></a>
<a href='https://coveralls.io/r/wollmers/LCS?branch=master'><img src='https://coveralls.io/repos/wollmers/LCS/badge.png?branch=master' alt='Coverage Status' /></a>
<a href='http://cpants.cpanauthors.org/dist/LCS'><img src='http://cpants.cpanauthors.org/dist/LCS.png' alt='Kwalitee Score' /></a>
<a href="http://badge.fury.io/pl/LCS"><img src="https://badge.fury.io/pl/LCS.svg" alt="CPAN version" height="18"></a>

=end html

=head1 SYNOPSIS

  use LCS;
  my $lcs = LCS->LCS( [qw(a b)], [qw(a b b)] );

  # $lcs now contains an arrayref of matching positions
  # same as
  $lcs = [
    [ 0, 0 ],
    [ 1, 2 ]
  ];

  my $all_lcs = LCS->allLCS( [qw(a b)], [qw(a b b)] );

  # same as
  $all_lcs = [
    [
      [ 0, 0 ],
      [ 1, 1 ]
    ],
    [
      [ 0, 0 ],
      [ 1, 2 ]
    ]
  ];

=head1 DESCRIPTION

LCS is an implementation based on the traditional LCS algorithm.

It contains reference implementations working slow but correct.

Also some utility methods are added to reformat the result.

=head2 CONSTRUCTOR

=over 4

=item new()

Creates a new object which maintains internal storage areas
for the LCS computation.  Use one of these per concurrent
LCS() call.

=back

=head2 METHODS

=over 4


=item LCS(\@a,\@b)

Finds a Longest Common Subsequence, taking two arrayrefs as method
arguments. It returns an array reference of corresponding
indices, which are represented by 2-element array refs.

  # position  0 1 2
  my $a = [qw(a b  )];
  my $b = [qw(a b b)];

  my $lcs = LCS->LCS($a,$b);

  # same like
  $lcs = [
      [ 0, 0 ],
      [ 1, 1 ]
  ];

=item LLCS(\@a,\@b)

Calculates the length of the Longest Common Subsequence.

  my $llcs = LCS->LLCS( [qw(a b)], [qw(a b b)] );
  print $llcs,"\n";   # prints 2

  # is the same as
  $llcs = scalar @{LCS->LCS( [qw(a b)], [qw(a b b)] )};

=item allLCS(\@a,\@b)

Finds all Longest Common Subsequences. It returns an array reference of all
LCS.

  my $all_lcs = LCS->allLCS( [qw(a b)], [qw(a b b)] );

  # same as
  $all_lcs = [
    [
      [ 0, 0 ],
      [ 1, 1 ]
    ],
    [
      [ 0, 0 ],
      [ 1, 2 ]
    ]
  ];

The purpose is mainly for testing LCS algorithms, as they only return one of the optimal
solutions. If we want to know, that the result is one of the optimal solutions, we need
to test, if the solution is part of all optimal LCS:

  use Test::More;
  use Test::Deep;
  use LCS;
  use LCS::Tiny;

  cmp_deeply(
    LCS::Tiny->LCS(\@a,\@b),
    any(@{LCS->allLCS(\@a,\@b)} ),
    "Tiny::LCS $a, $b"
  );

=item lcs2align(\@a,\@b,$LCS)

Returns the two sequences aligned, missing positions are represented as empty strings.

  use Data::Dumper;
  use LCS;
  print Dumper(
    LCS->lcs2align(
      [qw(a   b)],
      [qw(a b b)],
      LCS->LCS([qw(a b)],[qw(a b b)])
    )
  );
  # prints

  $VAR1 = [
            [
              'a',
              'a'
            ],
            [
              '',
              'b'
            ],
            [
              'b',
              'b'
            ]
  ];

=item align(\@a,\@b)

Returns the same as lcs2align() via calling LCS() itself.

=item sequences2hunks($a, $b)

Transforms two array references of scalars to an array of hunks (two element arrays).

=item hunks2sequences($hunks)

Transforms an array of hunks to two arrays of scalars.

  use Data::Dumper;
  use LCS;
  print Dumper(
    LCS->hunks2sequences(
      LCS->LCS([qw(a b)],[qw(a b b)])
    )
  );
  # prints (reformatted)
  $VAR1 = [ 0, 1 ];
  $VAR2 = [ 0, 2 ];


=item align2strings($hunks, $gap_character)

Returns two strings aligned with gap characters. The default gap character is '_'.

  use Data::Dumper;
  use LCS;
  print Dumper(
    LCS->align2strings(
      LCS->lcs2align([qw(a b)],[qw(a b b)],LCS->LCS([qw(a b)],[qw(a b b)]))
    )
  );
  $VAR1 = 'a_b';
  $VAR2 = 'abb';


=item fill_strings($string1, $string2, $fill_character)

Returns both strings filling up the shorter with $fill_character to the same length.

The default $fill_character is '_'.

=item clcs2lcs($compact_lcs)

Convert compact LCS to LCS.

=item lcs2clcs($compact_lcs)

Convert LCS to compact LCS.

=item max($i, $j)

Returns the maximum of two numbers.

=back

=head2 EXPORT

None by design.

=head1 STABILITY

Until release of version 1.00 the included methods, names of methods and their
interfaces are subject to change.

Beginning with version 1.00 the specification will be stable, i.e. not changed between
major versions.

=head1 REFERENCES

Ronald I. Greenberg. Fast and Simple Computation of All Longest Common Subsequences,
http://arxiv.org/pdf/cs/0211001.pdf

Robert A. Wagner and Michael J. Fischer. The string-to-string correction problem.
Journal of the ACM, 21(1):168-173, 1974.


=head1 SOURCE REPOSITORY

L<http://github.com/wollmers/LCS>

=head1 AUTHOR

Helmut Wollmersdorfer E<lt>helmut.wollmersdorfer@gmail.comE<gt>

=begin html

<a href='http://cpants.cpanauthors.org/author/wollmers'><img src='http://cpants.cpanauthors.org/author/wollmers.png' alt='Kwalitee Score' /></a>

=end html

=head1 COPYRIGHT

Copyright 2014- Helmut Wollmersdorfer

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut
